Mathlib/LinearAlgebra/FreeModule/Finite/Quotient.lean

```diff
@@ -0,0 +1,122 @@
+by principal ideals.
+-/
+noncomputable def quotientEquivPiSpan (N : Submodule R M) (b : Basis ι R M)
+    (h : Module.finrank R N = Module.finrank R M) :
+    (M ⧸ N) ≃ₗ[R] Π i, R ⧸ Ideal.span ({smithNormalFormCoeffs b h i} : Set R) := by
+  haveI := Fintype.ofFinite ι
+  -- Choose `e : M ≃ₗ N` and a basis `b'` for `M` that turns the map
+  -- `f := ((Submodule.subtype N).comp e` into a diagonal matrix:
+  -- there is an `a : ι → ℤ` such that `f (b' i) = a i • b' i`.
+  let a := smithNormalFormCoeffs b h
+  let b' := smithNormalFormTopBasis b h
+  let ab := smithNormalFormBotBasis b h
+  have ab_eq := smithNormalFormBotBasis_def b h
+  have mem_I_iff : ∀ x, x ∈ N ↔ ∀ i, a i ∣ b'.repr x i := by
+    intro x
+    simp_rw [ab.mem_submodule_iff', ab, ab_eq]
+    have : ∀ (c : ι → R) (i), b'.repr (∑ j : ι, c j • a j • b' j) i = a i * c i := by
+      intro c i
+      simp only [← MulAction.mul_smul, b'.repr_sum_self, mul_comm]
+    constructor
+    · rintro ⟨c, rfl⟩ i
+      exact ⟨c i, this c i⟩
+    · rintro ha
+      choose c hc using ha
+      exact ⟨c, b'.ext_elem fun i => Eq.trans (hc i) (this c i).symm⟩
+  -- Now we map everything through the linear equiv `M ≃ₗ (ι → R)`,
+  -- which maps `N` to `N' := Π i, a i ℤ`.
+  let N' : Submodule R (ι → R) := Submodule.pi Set.univ fun i => span R ({a i} : Set R)
+  have : Submodule.map (b'.equivFun : M →ₗ[R] ι → R) N = N' := by
+    ext x
+    simp only [N', Submodule.mem_map, Submodule.mem_pi, mem_span_singleton, Set.mem_univ,
+      Submodule.restrictScalars_mem, mem_I_iff, smul_eq_mul, forall_true_left, LinearEquiv.coe_coe,
+      Basis.equivFun_apply, mul_comm _ (a _), eq_comm (b := (x _))]
+    constructor
+    · rintro ⟨y, hy, rfl⟩ i
+      exact hy i
+    · rintro hdvd
+      refine ⟨∑ i, x i • b' i, fun i => ?_, ?_⟩ <;> rw [b'.repr_sum_self]
+      · exact hdvd i
+  refine (Submodule.Quotient.equiv N N' b'.equivFun this).trans (re₂₃ := inferInstance)
+    (re₃₂ := inferInstance) ?_
+  classical
+    let this :=
+      Submodule.quotientPi (show _ → Submodule R R from fun i => span R ({a i} : Set R))
+    exact this
+
+/--
+Quotients by submodules of full rank of free finite `ℤ`-modules are isomorphic
+to a direct product of `ZMod`.
+-/
+noncomputable def quotientEquivPiZMod (N : Submodule ℤ M) (b : Basis ι ℤ M)
+    (h : Module.finrank ℤ N = Module.finrank ℤ M) :
+    M ⧸ N ≃+ Π i, ZMod (smithNormalFormCoeffs b h i).natAbs :=
+  let a := smithNormalFormCoeffs b h
+  let e := N.quotientEquivPiSpan b h
+  let e' : (∀ i : ι, ℤ ⧸ Ideal.span ({a i} : Set ℤ)) ≃+ ∀ i : ι, ZMod (a i).natAbs :=
+    AddEquiv.piCongrRight fun i => ↑(Int.quotientSpanEquivZMod (a i))
+  (↑(e : (M ⧸ N) ≃ₗ[ℤ] _) : M ⧸ N ≃+ _).trans e'
+
+/-- A submodule of full rank of a free finite `ℤ`-module has a finite quotient.
+
+Can't be an instance because of the side condition and more importantly,
+because the choice of `Fintype` instance is non-canonical.
```

Comment:
But you could phrase the conclusion as `Finite (M ⧸ N)` instead, right?

Reply:
Sure, I just copied the original statement of [Ideal.fintypeQuotientOfFreeOfNeBot](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/FreeModule/IdealQuotient.html#Ideal.fintypeQuotientOfFreeOfNeBot). So I guess it would make more sense to change both of them. In which case, should I do it in this PR or maybe it's better to do that in a new PR?

Reply:
A new PR is fine with me.